src/main/java/de/bonndan/nivio/assessment/kpi/CustomKPI.java

```diff
@@ -0,0 +1,138 @@
+     */
+    @JsonCreator
+    public CustomKPI(@NonNull @JsonProperty(value = "label", required = true) String label,
+                     @Nullable @JsonProperty("messageLabel") String messageLabel,
+                     @Nullable @JsonProperty("ranges") Map<Status, String> ranges,
+                     @Nullable @JsonProperty("matches") Map<Status, String> matches
+    ) {
+        super(component -> {
+                    if (component instanceof Labeled) {
+                        return ((Labeled) component).getLabel(label);
+                    }
+                    throw new RuntimeException("Custom KPIs can only evaluate labels (custom fields).");
+                },
+                component -> {
+                    if (component instanceof Labeled) {
+                        return ((Labeled) component).getLabel(messageLabel);
+                    }
+                    return null;
+                });
+
+        this.label = label;
+        this.ranges = asRanges(ranges);
+        addSpecs(matches);
+    }
+
+    @Override
+    protected List<StatusValue> getStatusValues(String value, String message) {
+
+        List<StatusValue> values = new ArrayList<>();
+        for (Status status : Status.values()) {
+            Optional<Status> statusByRange = getStatusByRange(value);
+            if (statusByRange.isPresent()) {
+                values.add(new StatusValue(label, statusByRange.get(), message));
+                break;
+            }
+
+            if (!matches.containsKey(status)) {
+                continue;
+            }
+            boolean anyMatch = matches.get(status).stream().anyMatch(stringBooleanFunction -> stringBooleanFunction.apply(value));
+            if (anyMatch) {
+                values.add(new StatusValue(label, status, message));
+                break;
+            }
+        }
+
+        return values;
+    }
+
+    private Map<Status, Range<Double>> asRanges(Map<Status, String> ranges) {
+        Map<Status, Range<Double>> rangeMap = new HashMap<>();
+        if (ranges == null) {
+            return rangeMap;
+        }
+
+        ranges.forEach((status, s) -> {
+            String[] split;
+            if (s.contains(SEPARATOR)) {
+                split = s.split(SEPARATOR);
+            } else {
+                split = new String[2];
+                split[0] = s;
+                split[1] = s;
+            }
+            try {
+                rangeMap.put(status, Range.between(Double.valueOf(split[0]), Double.valueOf(split[1])));
+            } catch (NumberFormatException e) {
+                throw new ProcessingException("Failed to parse KPI range: " + s, e);
+            }
+        });
+        return rangeMap;
+    }
+
+    private Optional<Status> getStatusByRange(String value) {
+        if (value == null) {
+            return Optional.empty();
+        }
+
+        double d;
+        try {
+            d = Double.parseDouble(value);
+        } catch (NumberFormatException ignored) {
```

Comment:
Is there a valid case where it is ok to ignore this exception? Doesn't it hide more errors?

Reply:
We do not know the value to assess, which can be anything and can be supplied even by a third party system. So first we try parsing numbers to Double, and it that fails we assume it is a string and is compared against the regex matchers.

One improvement could be throwing an exception if only numeric ranges are defined for the KPI, but then we'd have to handle the exception, and a wrong value from a third party system is not necessarily exceptional.